### ISSUE TITLE
Configs: Enable no-undef by default

### ISSUE THREAD

**Issue by ufukbakan** [typescript-eslint/typescript-eslint#8825]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

"no-undef" rule should be enabled by default. Even this single line piece of shit is not caught by current recommended preset:
```ts
export default undefinedFunction();
```

### Impacted Configurations

@typescript-eslint/recommended

**Comment by bradzacher** [comment]
It's not caught by ESLint because it's caught by TS.
That is why it's turned off in our recommended set - to reduce noise from duplicated errors.

**Comment by ufukbakan** [comment]
@bradzacher Which command pipeline do you suggets? eslint -> tsc ?

I don't think its nice to allow this line to pass linting stage. I prefer rejecting this line even in commit stage via lint-staged & husky.

I understand, you don't want duplicate errors but I think pushing this code will just waste shared cloud resources. Anyways i will manually turn on this rule.

**Comment by bradzacher** [comment]
> Which command pipeline do you suggets? eslint -> tsc ?

All of our ESLint rules assume that you have well-formed, typecheck correct code.
Especially if you're using type-aware rules which are driven by your types.

So if you're running ESLint on half-finished code or code with errors then you will likely have other noise from ESLint - essentially you're entering "undefined behaviour" for a lot of our rules.

Additionally if you're relying on project references and such - getting correct type-aware linting relies on the `.d.ts` output of referenced projects - i.e. again your TS build needs to pass for the lint run to work as expected.

> pushing this code will just waste shared cloud resources

Correct - which is why you should typecheck your code before you push it to avoid running a CI step that's guaranteed to fail.
Your IDE will tell you that you have a type-error as you're writing the code so such an error will be highlighted for you live.

**Comment by ufukbakan** [comment]
what you mean by "you should typecheck your code before you push it"
isn't eslint for it?
do you mean manuel typechecking by oversighting text seen on the IDE?
how do you do typecheck in an automated code pipeline


> > Which command pipeline do you suggets? eslint -> tsc ?
> 
> All of our ESLint rules assume that you have well-formed, typecheck correct code. Especially if you're using type-aware rules which are driven by your types.
> 
> So if you're running ESLint on half-finished code or code with errors then you will likely have other noise from ESLint - essentially you're entering "undefined behaviour" for a lot of our rules.
> 
> Additionally if you're relying on project references and such - getting correct type-aware linting relies on the `.d.ts` output of referenced projects - i.e. again your TS build needs to pass for the lint run to work as expected.
> 
> > pushing this code will just waste shared cloud resources
> 
> Correct - which is why you should typecheck your code before you push it to avoid running a CI step that's guaranteed to fail. Your IDE will tell you that you have a type-error as you're writing the code so such an error will be highlighted for you live.

**Comment by bradzacher** [comment]
I mean use typescript?
If you're writing typescript code then typescript is checking your code.

So you run `tsc` from your local CLI and TS will typecheck your code and report errors.

ESLint is not a typechecker - it is a linter and our project is designed to augment / build on top of the checks that TS does.

**Comment by ufukbakan** [comment]
Thanks, so you suggest using tsc before linting in development environment.
Then we can safely run only tsc in "build runner / production envrionment" right?
